### ISSUE TITLE
`environment.binsh` does no sanity checking

### ISSUE THREAD

**Issue by tejing1** [NixOS/nixpkgs#270978]
### Describe the bug
A user I was interacting with on discord set `environment.binsh = pkgs.fish;` in an attempt to change their login shell. Naturally, the world immediately blew up. It wasn't even pointing at the binary, to be clear, but the root directory of the store object. They were lead down this path partly by the nixos wiki, which has already been changed to not even list the option on that page.

### Steps To Reproduce
1. set `environment.binsh = pkgs.fish;` in `configuration.nix`
2. rebuild
3. watch the fireworks

### Expected behavior
While such "skill issue" problems cannot be solved in any particular totality, given the consequences of a bad value here, I do think it's reasonable for the `binsh` activation script to check that the link target is a runnable program that seems, on cursory examination, to act like a posix-compliant `sh`. If it does not seem to be, and there is an existing link, it could print a conspicuous error message and leave the existing link in place.

Since some valid targets may examine arg0, we may need to alter the temporary link to sit inside a hidden directory or the like, when examining behavior.

Even just checking that it is in fact an executable file would be a big improvement.

### Priorities

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

**Comment by Aleksanaa** [comment]
https://github.com/NixOS/nixpkgs/blob/8abf7c04993d786f5561d1c3ce35d0ebbc61ad1d/nixos/modules/config/shells-environment.nix#L142

These options exist to make the system structure clearer and easier to maintain, not for users to modify according to their own ideas. If you really have a good reason to touch them, you should be aware of what you are doing.

**Comment by NobbZ** [comment]
@Aleksanaa True, its not visible. Still, **other** options refer to it, by mentioning it as their default.

`systemd.services.<name>.confinement.binSh`

This makes the information leak obviously. And in the case Teijing mentions, an unexperienced user found a mention of that value in the (inofficial) wiki, missed or ignored the warning to not touch the option, and set it to `pkgs.fish`, which blew up their system, as from then on `/bin/sh` was a link to a *folder*. Subsequent activations failed because of that, as the "fixed" link was tried to be created at `/bin/sh/sh` due to the way how everything is implemented.

After the problem and the source of the information got identified, some kind member quickly removed that section from the Wiki, though god knows how many blog posts out there some how mention this option as the way to go? There we can not just edit the text. God knows how many AIs have or will suggest to set this value to garbage?

Documented or not, this is an option that can cause harm to the system in a way that even experienced users might fail to resolve.

A sanity check makes a lot of sense.
